### PR TITLE
feat: interactive authentication through OAuth2 

### PR DESCRIPTION
**Pull request by zepatrik**
courtesy of @alnr
closes #357

### CODE REVIEW

**Review by jonas-jonas — CHANGES_REQUESTED** [review]

File: cmd/cloudx/client/api_key.go [hunk 1]
```diff
@@ -5,40 +5,100 @@ package client
...
+		strings.NewReader(fmt.Sprintf(`{"name":"%s"}`, name)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	res, err := c.Do(req)
+	if err != nil {
+		return nil, err
+	}
```
Comment (alnr) [inline cmd/cloudx/client/api_key.go:66]:
```suggestion
	}
	defer res.Body.Close()
```

File: cmd/cloudx/client/api_key.go [hunk 2]
```diff
@@ -5,40 +5,100 @@ package client
...
+		return err
+	}
+	req, err := http.NewRequest(http.MethodDelete, fmt.Sprintf("%s/workspaces/%s/tokens/%s", baseURL.String(), workspaceID, keyID), nil)
+	if err != nil {
+		return err
+	}
+	res, err := c.Do(req)
+	if err != nil {
+		return err
+	}
```
Comment (alnr) [inline cmd/cloudx/client/api_key.go:95]:
```suggestion
	}
	}
	defer res.Body.Close()
```

File: cmd/cloudx/client/command_helper.go [hunk 3]
```diff
@@ -142,29 +171,42 @@ func NewCommandHelper(ctx context.Context, opts ...CommandHelperOption) (*Comman
 		noConfirm:        false,
 		VerboseErrWriter: io.Discard,
 		Stdin:            bufio.NewReader(os.Stdin),
-		pwReader: func() ([]byte, error) {
-			return term.ReadPassword(int(os.Stdin.Fd()))
+		openBrowserHook: func(uri string) error {
```
Comment (alnr) [inline cmd/cloudx/client/command_helper.go:174]:
```suggestion
		openBrowserHook: func(uri string) {
```

File: cmd/cloudx/client/command_helper_test.go [hunk 4]
```diff
@@ -28,64 +35,92 @@ import (
 var updatedProjectConfig json.RawMessage
 
 func TestMain(m *testing.M) {
-	testhelpers.RunAgainstStaging(m)
+	if err := playwright.Install(&playwright.RunOptions{
+		Browsers: []string{"chromium"},
+	}); err != nil {
+		panic(err)
+	}
+	os.Exit(m.Run())
```
Comment (alnr) [inline cmd/cloudx/client/command_helper_test.go:43]:
```suggestion
	m.Run()
```

`TestMain` does the correct thing on its own.

**Comment by alnr** [comment]
There are a couple of places where `:443` is appended to URLs. Those need to go because it interferes with the OAuth2 URL whitelisting.

File: cmd/cloudx/client/command_helper.go (hunk 3 above)
Comment (zepatrik) [reply cmd/cloudx/client/command_helper.go:174]: We need the error for testing unfortunately...

File: cmd/cloudx/client/api_key.go (hunk 2 above)
Comment (zepatrik) [reply cmd/cloudx/client/api_key.go:95]: Ideally we'll not merge this, but have the SDK published. I'll keep it open as a reminder.

File: cmd/cloudx/client/command_helper_test.go (hunk 4 above)
Comment (zepatrik) [reply cmd/cloudx/client/command_helper_test.go:43]: Interesting, didn't know they eventually fixed this. It is already for quite some time, appareltly go1.15 https://github.com/golang/go/commit/2f54081adfc967836842c96619d241378400ece6

File: cmd/cloudx/client/sdks.go [hunk 5]
```diff
@@ -22,58 +26,105 @@ const (
...
+	} else if h.sessionToken != nil {
+		tokenSource = oauth2.StaticTokenSource(&oauth2.Token{AccessToken: *h.sessionToken})
+		baseURL = CloudConsoleURL
+	} else {
+		config, err := h.GetAuthenticatedConfig(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		tokenSource = config.TokenSource(ctx)
+		baseURL = CloudConsoleURL
```
[the remark this answers is not in the document]
Comment (zepatrik) [reply cmd/cloudx/client/sdks.go:118]: Good catch...

File: cmd/cloudx/relationtuples/relationtuples.go [hunk 6]
```diff
@@ -68,15 +68,20 @@ func forwardConnectionInfo(cmd *cobra.Command) {
 			return err
 		}
 
-		key, err := h.CreateAPIKey(ctx, project.Id, "keto-temp-"+randx.MustString(8, randx.AlphaNum))
+		key, err := h.CreateProjectAPIKey(ctx, project.Id, "keto-temp-"+randx.MustString(8, randx.AlphaNum))
 		if err != nil {
 			return err
 		}
-		defer func() { _ = h.DeleteAPIKey(ctx, project.Id, key.Id) }()
+		defer func() { _ = h.DeleteProjectAPIKey(ctx, project.Id, key.Id) }()
```
Comment (alnr) [inline cmd/cloudx/relationtuples/relationtuples.go:75]: Keto should also use OAuth2. I can do that in a follow-up.

File: cmd/cloudx/testhelpers/testhelpers.go [hunk 7]
```diff
@@ -64,96 +64,45 @@ func ReadConfig(t testing.TB, configDir string) *client.Config {
...
+	return client.ContextWithOptions(ctx, client.WithConfigLocation(NewConfigFile(t)))
+}
 
-	c, err := client.NewOryProjectClient()
+func WithDuplicatedConfigFile(ctx context.Context, t testing.TB, originalFile string) context.Context {
+	dst, err := os.Create(NewConfigFile(t))
 	require.NoError(t, err)
-
-	res, _, err := c.FrontendAPI.ToSession(context.Background()).XSessionToken(ac.SessionToken).Execute()
+	src, err := os.Open(originalFile)
```
Comment (alnr) [inline cmd/cloudx/testhelpers/testhelpers.go:85]: must be closed

File: cmd/cloudx/relationtuples/relationtuples.go (hunk 6 above)
Comment (zepatrik) [reply cmd/cloudx/relationtuples/relationtuples.go:75]: True, let's do that in a follow-up.

**Pull request merged**